marlowe-chain-sync/libchainsync/Language/Marlowe/Runtime/ChainSync/NodeClient.hs

```diff
@@ -0,0 +1,272 @@
+  where
+    queryToJSON :: Q.Some (QueryInMode CardanoMode) -> Value
+    queryToJSON (Q.Some query) = String . T.pack $ show query
+
+
+nodeClient :: Component IO (NodeClientDependencies r) NodeClient
+nodeClient =
+  component \NodeClientDependencies{..} ->
+    do
+      queryChannel <- newTChan
+      submitChannel <- newTChan
+      let
+        runNodeClient = mask \restore ->
+          do
+            ev <- newEvent eventBackend Connect
+            result <-
+              try @SomeException
+                . restore
+                $ connectToLocalNode LocalNodeClientProtocols
+                  {
+                    -- We don't use the block and tip information, but the chain
+                    -- sync client keeps the connection open. Without this client,
+                    -- the node would close the connection in about three seconds,
+                    -- making it impossible to submit and query.
+                    localChainSyncClient = LocalChainSyncClient chainSyncClient
+                  , localTxSubmissionClient = Just $ submitClient submitChannel
+                  , localTxMonitoringClient = Nothing
+                  , localStateQueryClient = Just $ queryClient queryChannel
+                  }
+            case result of
+              Left ex -> failEvent ev ex *> throw ex
+              Right _ -> pure ()
+      pure
+        (
+          runNodeClient
+        , NodeClient
+          {
+            submitTxToNode = submitTxToNodeChannel eventBackend submitChannel
+          , queryNode = queryNodeChannel eventBackend queryChannel
+          }
+        )
+
+
+type SubmitChannel = TChan SubmitJob
+
+
+data SubmitJob =
+  SubmitJob
+  {
+    tx :: TxInMode CardanoMode
+  , submitResultTMVar :: TMVar (S.SubmitResult (TxValidationErrorInMode CardanoMode))
+  }
+
+
+submitTxToNodeChannel
+  :: EventBackend IO r NodeClientSelector
+  -> SubmitChannel
+  -> TxInMode CardanoMode
+  -> IO (S.SubmitResult (TxValidationErrorInMode CardanoMode))
+submitTxToNodeChannel eventBackend channel tx =
+  withEvent eventBackend Submit \event ->
+    do
+      event `addField` tx
+      submitResultTMVar <- newEmptyTMVarIO
+      atomically $ writeTChan channel SubmitJob{..}
+      atomically $ takeTMVar submitResultTMVar
+
+
+submitClient
+  :: SubmitChannel
+  -> S.LocalTxSubmissionClient (TxInMode CardanoMode) (TxValidationErrorInMode CardanoMode) IO ()
+submitClient channel =
+  let
+    next =
+      do
+        SubmitJob{..} <- atomically $ readTChan channel
+        pure . S.SendMsgSubmitTx tx
+          $ \result ->
+            do
+              atomically $ putTMVar submitResultTMVar result
+              next
+  in
+    S.LocalTxSubmissionClient next
+
+
+
+type QueryChannel = TChan (Q.Some QueryJob)
+
+
+data QueryJob result =
+  QueryJob
+  {
+    point :: Maybe ChainPoint
+  , query :: QueryInMode CardanoMode result
+  , queryResultTMVar :: TMVar (Either Q.AcquireFailure result)
+  }
```

Comment:
Rather than adding the additional `Some` wrapper, you could also existentially quantify `result` within the `QueryJob` constructor

```suggestion
data QueryJob = forall result.
  QueryJob
  {
    point :: Maybe ChainPoint
  , query :: QueryInMode CardanoMode result
  , queryResultTMVar :: TMVar (Either Q.AcquireFailure result)
  }
```

Reply:
Nice! Thanks, fixed in 847d17a911fd63d51230b1c439eb1ceb0d966f5c.